text/0000-process-type-arguments.md

```diff
@@ -0,0 +1,207 @@
+
+Instead, at execution time, the lifecycle should allow a user to specify the arguments to append to an existing process type.
+
+Feedback from users in buildpacks slack and in issues, indicates that should in fact be the default behavior:
+- https://github.com/buildpacks/lifecycle/issues/280
+- https://github.com/buildpacks/pack/issues/468#issuecomment-593139689
+
+## Problem 2: Arguments for Shell Processes
+Currently, arguments defined in a shell process are treated as arguments to bash, rather than as arguments to the command.
+Given a process:
+```
+[[process]]
+command = "<CMD>"
+args = ["ARG1", "ARG2"]
+```
+the `launcher` currently executes `bash -c "<CMD>" "ARG1" "ARG2"`. This makes it hard to use arguments intuitively in a shell process.
+For example, the following does not print `hello world`.
+```
+[[process]]
+command = "echo"
+args = ["hello", "world"]
+```
+Instead, users must construct a string command that includes arguments
+```
+[[process]]
+command = "echo hello world"
+args = []
+```
+This already causes confusion for users (https://github.com/buildpacks/lifecycle/issues/281). The problem is compounded if we allow users to supply additional arguments. We cannot reliably append arguments to the `command` string without introducing tokenization or shell parsing errors. Constructing a `command` that accepts args in the way they are currently provided to `bash` is confusing and difficult (buildpack authors likely will not do this correctly).
+
+This RFC proposes changes to the `launcher`'s handling of args to make the execution of the following process
+```
+[[process]]
+command = "<CMD>"
+args = ["ARG1", "ARG2"]
+```
+produce results equivalent to the execution of `"<CMD>" "<ARG1>" "<ARG2>"` in a shell
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Terminology
+* shell process - a process with `direct=false`
+
+## Philosophy
+The goal is to make the runtime usage of a CNB container as similar as possible to the runtime usage on a non-CNB container, without removing functionality or violating assumptions that have been firmly established in the buildpack community (e.g. defaulting to `web`).
+
+Because the `launcher` must occupy the container `entrypoint`, we will instead use `CNB_PROCESS_TYPE` as a CNB-specific analog to the `entrypoint`. Users can already set `CNB_PROCESS_TYPE` to toggle between processes. Setting `CNB_PROCESS_TYPE=override` will be analogous to clearing the `entrypoint`.
+
+## Behavior
+### Selecting a process type at runtime
+Users will no longer be able to pass a process-type to the launcher as a positional argument. Instead, they must set `CNB_PROCESS_TYPE` in the running container to select a process type.
```

Comment:
I would rather see us *require* the process name in the CLI. For an image with a `web` process only, this would yield the following cases:

* `docker run <image>` - works as it does today
* `docker run <image> web` - works as it does today
* `docker run <image> web asdf` - passes `["asdf"]` as args
* `docker run <image> bash` - fails to locate a process named "bash" and sets `command = "bash"` instead
* `docker run <image> -- bash` - fails to locate a process named "bash" and sets `command = "bash"` and `direct = true`

Reply:
If we went this route I don't think process type lookup should be attempted if the first arg is `--`. `--` means `direct` and we should not change the value of `direct` for a buildpack provided process. If `--` is provided it is always a custom command

Reply:
I feel pretty strongly that this makes our images less compliant / normal compared to images built with Dockerfiles. Most app/service images I can find allow you to provide args. K8s calls the field `args`, and the assumption that entrypoint == `command` is built-in to k8s: https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/

I feel like nearly every K8s user is going to be confused by this.

Reply:
But the args we're talking about aren't args to the launcher. They are args to the process it's run. I feel like it makes sense to always include the process name in what we pass to launcher.

This is the way it works today. Do we have feedback today that it confusing k8s users?